### DISCUSSION TITLE
RLS and "groups" of users

### DISCUSSION THREAD

**Discussion by Towerful** [supabase/supabase#6727] [Questions] [answered]
Say I had a team of users, and some tables have a "team_id" column.   
A user cannot belong to multiple teams (I'm just using teams as a stand in for this...)   
Using RLS, how do I go about securing this?  

I presume this is what the User `user_metadata` is about, as this is a service-account level property and isn't user editable.  
Is there a way to access this via the policies `auth` helpers?  
https://supabase.com/docs/learn/auth-deep-dive/auth-policies says  

> Supabase comes with three built-in helper functions: auth.email(), auth.uid() and auth.role().

which suggests I cannot access the metadata.  

-----------------

Ok, looking at the SQL Snippets on cloud supabase, and the db init scripts in the docker folder, and this issue https://github.com/supabase/supabase/discussions/1148 
I am coming up with this (and this is a guess, I'm new to all this!): 
```pgsql
create or replace function auth.team_id() 
returns bigint
language sql stable
as $$
  select 
    ((current_setting('request.jwt.claims')::jsonb ->> 'user_metadata')::jsonb ->> 'team_id'),'')::bigint
$$;
```

and using it as

```pgsql
create policy "Allow team members access to their discussions" on public.discussions
  for select using (auth.team_id() = team_id);
```

I'll be creating users via a service account:  
```js
  let { user, data } = await supabase.auth.api.createUser({
      email: "someone@example.com",
      password: "my test password",
      email_confirm: true,
      user_metadata: {
        team_id: 1
      }
    })
```

---------

as an aside, how do I go about testing & debugging complex functions and policies?

**Comment by GaryAustin1** [comment, marked as answer]
As commented in the link you used https://github.com/supabase/supabase/discussions/1148, app_metadata might be a better option as client code can change user_metadata if they have email/password permissions.

As far complex policies, putting them in functions makes them easier to test with the sql editor in the dashboard.  

If you want to test the policies on a table, this article goes thru one way to do that:
https://www.slip.so/tutorials/row-level-security-performance-in-supabase

You can use the postgres log to check for function errors and use raise log https://www.postgresql.org/docs/current/plpgsql-errors-and-messages.html to add debug info in functions.

**Reply by Towerful** [reply]
Haha, I was too busy thinking about how to implement it to realise that the issue I linked actually fully describes everything I need to do, and why to use app_metadata and even how to debug it.  
Sorry bout that.

That slip.so article is an interesting read, tho!

**Comment by Lock23Supply** [comment]
https://github.com/supabase-community/supabase-custom-claims
